# This code may be dead, I can't get a good r2 no matter what I do

In [1]:
#Dependencies
import pandas as pd
import keras_tuner as kt
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.callbacks import EarlyStopping

### Getting Data for feeding the model

In [2]:
#Read in cleaned csv into a pandas dataframe
filename = "training_data.csv"
try:
    df = pd.read_csv(f"../../Data/{filename}")
except FileNotFoundError:
    raise FileNotFoundError(f"The file '{filename}' does not exist. Please make sure it is in the 'Data' folder.")

df = df.dropna()
df.head()

Unnamed: 0  fips_code  year  niaaa_pop_21_plus  census_total_pop  \
0          51          1  2011            3467668           4802740   
1          52          2  2011             502699            722718   
2          53          4  2011            4574939           6482505   
3          54          5  2011            2103411           2937979   
4          55          6  2011           26698281          37691912   

   census_percent_pop_21_plus  census_percent_pop_18_24  \
0                       0.722                     0.101   
1                       0.696                     0.107   
2                       0.704                     0.100   
3                       0.714                     0.096   
4                       0.708                     0.106   

   census_percent_pop_25_34  census_percent_pop_35_44  \
0                     0.127                     0.128   
1                     0.147                     0.129   
2                     0.134                     0.127   
3                     0.130                     0.124   
4                     0.143                     0.137   

   census_percent_pop_45_54  ...  tpc_state_beer_tax_rates_dollar_gal  \
0                     0.143  ...                                 1.05   
1                     0.149  ...                                 1.07   
2                     0.130  ...                                 0.16   
3                     0.137  ...                                 0.24   
4                     0.139  ...                                 0.20   

   bea_personal_income_dollars  bea_disp_inc_dollars  \
0                     167882.2              152835.4   
1                      37870.6               34789.2   
2                     227203.3              206128.2   
3                     100449.5               91406.0   
4                    1713428.0             1504645.4   

   bea_percapita_personal_income_dollars  bea_percapita_disp_inc_dollars  \
0                                  34887                           31760   
1                                  52415                           48150   
2                                  35244                           31975   
3                                  34180                           31103   
4                                  45509                           39963   

   census_median_household_inc_dollars  niaaa_total_beer_vol_consumed_gallons  \
0                                41415                              101246000   
1                                67825                               14688320   
2                                46709                              132997688   
3                                38758                               51707608   
4                                57287                              643901979   

   niaaa_legal_adult_per_capita_beer_consumed_gallons  \
0                                              29.20    
1                                              29.22    
2                                              29.07    
3                                              24.58    
4                                              24.12    

   ba_craft_beer_produced_gallons  \
0                         29202.0   
1                       4870317.0   
2                       2141480.0   
3                        126449.0   
4                      62627626.0   

   ba_legal_adult_per_capita_craft_beer_produced_gallons  
0                                               0.01      
1                                               9.69      
2                                               0.47      
3                                               0.06      
4                                               2.35      

[5 rows x 39 columns]

In [3]:
#Split data for testing and training
yColName = "niaaa_total_beer_vol_consumed_gallons"
X = df.drop(yColName, axis=1).values
y = df[yColName].values

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [5]:
#Scale the data

#Create a StandardScaler instances
scaler = StandardScaler()

#Fit the StandardScaler
X_scaler = scaler.fit(X_train)

#Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

### Model Creation

In [6]:
#Create model function for keras tuner
regression = True
MAX_NEURONS = 128
MAX_LAYERS = 10

def create_model(hp):
    nnModel = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation', ['relu', 'tanh', 'sigmoid', 'linear'])

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, MAX_LAYERS)):
        nnModel.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i), min_value=1, max_value=MAX_NEURONS, step=8), activation=activation))

    # Last layer for binary classification
    if not regression:  
        nnModel.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))
        nnModel.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    elif regression:
        nnModel.add(tf.keras.layers.Dense(units=1, activation="linear"))
        nnModel.compile(loss='mean_squared_error', optimizer='adam', metrics=['mse'])

    return nnModel

In [7]:
#Use the tuner to generate the best model
tuner = kt.Hyperband(
    create_model,
    objective="mse",
    max_epochs=20,
    hyperband_iterations=2
)

tuner.search(X_train_scaled, y_train, epochs=20, validation_data=(X_test_scaled, y_test))
nn = tuner.get_best_models(num_models=1)[0]

Trial 60 Complete [00h 00m 05s]
mse: 3.2354467529621504e+16

Best mse So Far: 480095001116672.0
Total elapsed time: 00h 02m 58s


In [11]:
#Create an EarlyStopping callback
early_stopping = EarlyStopping(monitor='loss', patience=5, restore_best_weights=True)

fit_model = nn.fit(X_train_scaled, y_train, epochs=500, callbacks=[early_stopping])

Epoch 1/500
12/12 [==============================] - 0s 4ms/step - loss: 329961667821568.0000 - mse: 329961667821568.0000
Epoch 2/500
12/12 [==============================] - 0s 2ms/step - loss: 294206501289984.0000 - mse: 294206501289984.0000
Epoch 3/500
12/12 [==============================] - 0s 2ms/step - loss: 347410408669184.0000 - mse: 347410408669184.0000
Epoch 4/500
12/12 [==============================] - 0s 2ms/step - loss: 374719555567616.0000 - mse: 374719555567616.0000
Epoch 5/500
12/12 [==============================] - 0s 2ms/step - loss: 628592421634048.0000 - mse: 628592421634048.0000
Epoch 6/500
12/12 [==============================] - 0s 1ms/step - loss: 384115937378304.0000 - mse: 384115937378304.0000
Epoch 7/500
12/12 [==============================] - 0s 3ms/step - loss: 296424247918592.0000 - mse: 296424247918592.0000


In [12]:
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

4/4 - 0s - loss: 462938217381888.0000 - mse: 462938217381888.0000 - 23ms/epoch - 6ms/step
Loss: 462938217381888.0, Accuracy: 462938217381888.0


In [13]:
from sklearn.metrics import r2_score

# Assuming nn is your trained model and X_test_scaled, y_test are your test data
predictions = nn.predict(X_test_scaled)

# Flatten predictions and ground truth for compatibility with r2_score
predictions = predictions.flatten()

# Calculate R-squared
r_squared = r2_score(y_test, predictions)

print(f'R-squared: {r_squared}')

4/4 [==============================] - 0s 2ms/step
R-squared: 0.9824075037687696


#### All cells below are previous outputs of the cell above as we go through the process of optimization, along with what we changed

After many issues, the code finally works properly, resulting in an r2 of 0.9824. This is suspiciously high, so we are seeking other models to confirm.